File: 210-course-schedule-ii/210-course-schedule-ii.py
class Solution:
    def findOrder(self, numCourses: int, prerequisites: List[List[int]]) -> List[int]:
        ans, in_coming, out_going = [], [0] * numCourses, defaultdict(set)
        queue = deque()        
        
        for course, pre in prerequisites:
            out_going[pre].add(course)
            in_coming[course] += 1
        
        for i in range(numCourses):
            if in_coming[i] == 0:
                queue.append(i)
                
        while queue:
            pre = queue.popleft()
            ans.append(pre)
            
            for course in out_going[pre]:
                in_coming[course] -= 1
                if in_coming[course] == 0:
                    queue.append(course)
                    
        if len(ans) == numCourses:
            return ans
        else:
            return []
            
        

File: 210-course-schedule-ii/210-course-schedule-ii.java
class Solution {
    private final int WHITE = 1;
        private final int GREY = 2;
        private final int BLACK = 3;
        
        private Map<Integer, Integer> color;
        private Map<Integer, ArrayList<Integer>> adjList;
        private List<Integer> topologicalOrd;
        private boolean isPossible;
        
        private void init(int numCourses){
            isPossible = true;
            color = new HashMap<>();
            adjList = new HashMap<>();
            topologicalOrd = new ArrayList<>();
            
            for (int i = 0;  i < numCourses; i++){
                color.put(i, WHITE);
            }
        }
    
        private void dfs(int node){
            if (!isPossible){
                return;
            }
            
            color.put(node, GREY);
            for (int neighbour : adjList.getOrDefault(node, new ArrayList<Integer>())){
                if (color.get(neighbour) == WHITE){
                    dfs(neighbour);
                }else if (color.get(neighbour) == GREY){
                    isPossible = false;
                }
            }
            topologicalOrd.add(node);
             color.put(node, BLACK);
        }
    public int[] findOrder(int numCourses, int[][] prerequisites) {
        init(numCourses);
    
        for(int[] cor_pre: prerequisites){
            int course = cor_pre[0];
            int pre = cor_pre[1];
            ArrayList<Integer>lst = adjList.getOrDefault(pre, new ArrayList<Integer>());
            lst.add(course);
            adjList.put(pre, lst);
        }
            
        for(int i = 0; i < numCourses; i++){
            if( color.get(i) == WHITE){
                dfs(i);
            }
        }
        
        int [] ord;
        if (isPossible){
            ord = new int[numCourses];
            for(int i = 0; i < numCourses; ++i){
                ord[i] = topologicalOrd.get(numCourses - i - 1);
            }
        }else{
            ord = new int[0];
        }
        
        return ord;
    }
}